<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mikropml">
<title>mikropml: User-Friendly R Package for Supervised Machine Learning Pipelines • mikropml</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="mikropml: User-Friendly R Package for Supervised Machine Learning Pipelines">
<meta property="og:description" content="mikropml">
<meta property="og:image" content="http://www.schlosslab.org/mikropml/logo.png">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mikropml</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.6.0.9001</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Paper</h6>
    <a class="dropdown-item" href="../articles/paper.html">mikropml: User-Friendly R Package for Supervised Machine Learning Pipelines</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Vignettes</h6>
    <a class="dropdown-item" href="../articles/introduction.html">Introduction to mikropml</a>
    <a class="dropdown-item" href="../articles/preprocess.html">Preprocessing data</a>
    <a class="dropdown-item" href="../articles/tuning.html">Hyperparameter tuning</a>
    <a class="dropdown-item" href="../articles/parallel.html">Parallel processing</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SchlossLab/mikropml/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>mikropml: User-Friendly R Package for Supervised Machine Learning Pipelines</h1>
                        <h4 data-toc-skip class="author">Begüm D.
Topçuoğlu, Zena Lapp, Kelly L. Sovacool, Evan Snitkin, Jenna Wiens,
Patrick D. Schloss</h4>
            
            <h4 data-toc-skip class="date">2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SchlossLab/mikropml/blob/HEAD/vignettes/paper.Rmd" class="external-link"><code>vignettes/paper.Rmd</code></a></small>
      <div class="d-none name"><code>paper.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p><img src="mikropml-logo.png" width="120"></p>
<p>Machine learning (ML) for classification and prediction based on a
set of features is used to make decisions in healthcare, economics,
criminal justice and more. However, implementing an ML pipeline
including preprocessing, model selection, and evaluation can be
time-consuming, confusing, and difficult. Here, we present <a href="http://www.schlosslab.org/mikropml/"><code>mikropml</code></a>
(pronounced “meek-ROPE em el”), an easy-to-use R package that implements
ML pipelines using regression, support vector machines, decision trees,
random forest, or gradient-boosted trees. The package is available on <a href="https://github.com/SchlossLab/mikropml/" class="external-link">GitHub</a>, <a href="https://cran.r-project.org/package=mikropml" class="external-link">CRAN</a>, and <a href="https://anaconda.org/conda-forge/r-mikropml" class="external-link">conda</a>.</p>
</div>
<div class="section level2">
<h2 id="statement-of-need">Statement of need<a class="anchor" aria-label="anchor" href="#statement-of-need"></a>
</h2>
<p>Most applications of machine learning (ML) require reproducible steps
for data pre-processing, cross-validation, testing, model evaluation,
and often interpretation of why the model makes particular predictions.
Performing these steps is important, as failure to implement them can
result in incorrect and misleading results <span class="citation">(Teschendorff 2019; Wiens et al. 2019)</span>.</p>
<p>Supervised ML is widely used to recognize patterns in large datasets
and to make predictions about outcomes of interest. Several packages
including <code>caret</code> <span class="citation">(Kuhn 2008)</span>
and <code>tidymodels</code> <span class="citation">(Kuhn, Wickham, and
RStudio 2020)</span> in R, <code>scikitlearn</code> <span class="citation">(Pedregosa et al. 2011)</span> in Python, and the H2O
<code>autoML</code> platform <span class="citation">(H2O.ai 2020)</span>
allow scientists to train ML models with a variety of algorithms. While
these packages provide the tools necessary for each ML step, they do not
implement a complete ML pipeline according to good practices in the
literature. This makes it difficult for practitioners new to ML to
easily begin to perform ML analyses.</p>
<p>To enable a broader range of researchers to apply ML to their problem
domains, we created <a href="https://github.com/SchlossLab/mikropml/" class="external-link"><code>mikropml</code></a>,
an easy-to-use R package <span class="citation">(R Core Team
2020)</span> that implements the ML pipeline created by Topçuoğlu <em>et
al.</em> <span class="citation">(Topçuoğlu et al. 2020)</span> in a
single function that returns a trained model, model performance metrics
and feature importance. <code>mikropml</code> leverages the
<code>caret</code> package to support several ML algorithms: linear
regression, logistic regression, support vector machines with a radial
basis kernel, decision trees, random forest, and gradient boosted trees.
It incorporates good practices in ML training, testing, and model
evaluation <span class="citation">(Topçuoğlu et al. 2020; Teschendorff
2019)</span>. Furthermore, it provides data preprocessing steps based on
the FIDDLE (FlexIble Data-Driven pipeLinE) framework outlined in Tang
<em>et al.</em> <span class="citation">(Tang et al. 2020)</span> and
post-training permutation importance steps to estimate the importance of
each feature in the models trained <span class="citation">(Breiman 2001;
Fisher, Rudin, and Dominici 2018)</span>.</p>
<p><code>mikropml</code> can be used as a starting point in the
application of ML to datasets from many different fields. It has already
been applied to microbiome data to categorize patients with colorectal
cancer <span class="citation">(Topçuoğlu et al. 2020)</span>, to
identify differences in genomic and clinical features associated with
bacterial infections <span class="citation">(Lapp et al. 2020)</span>,
and to predict gender-based biases in academic publishing <span class="citation">(Hagan et al. 2020)</span>.</p>
</div>
<div class="section level2">
<h2 id="mikropml-package">mikropml package<a class="anchor" aria-label="anchor" href="#mikropml-package"></a>
</h2>
<p>The <code>mikropml</code> package includes functionality to
preprocess the data, train ML models, evaluate model performance, and
quantify feature importance (Figure 1). We also provide <a href="http://www.schlosslab.org/mikropml/articles/index.html">vignettes</a>
and an <a href="https://github.com/SchlossLab/mikropml-snakemake-workflow" class="external-link">example
Snakemake workflow</a> <span class="citation">(Köster and Rahmann
2012)</span> to showcase how to run an ideal ML pipeline with multiple
different train/test data splits. The results can be visualized using
helper functions that use <code>ggplot2</code> <span class="citation">(Wickham 2016)</span>.</p>
<p>While mikropml allows users to get started quickly and facilitates
reproducibility, it is not a replacement for understanding the ML
workflow which is still necessary when interpreting results <span class="citation">(Pollard et al. 2019)</span>. To facilitate
understanding and enable one to tailor the code to their application, we
have heavily commented the code and have provided supporting
documentation which can be read <a href="http://www.schlosslab.org/mikropml/">online</a>.</p>
<div class="section level3">
<h3 id="preprocessing-data">Preprocessing data<a class="anchor" aria-label="anchor" href="#preprocessing-data"></a>
</h3>
<p>We provide the function <code><a href="../reference/preprocess_data.html">preprocess_data()</a></code> to preprocess
features using several different functions from the <code>caret</code>
package. <code><a href="../reference/preprocess_data.html">preprocess_data()</a></code> takes continuous and categorical
data, re-factors categorical data into binary features, and provides
options to normalize continuous data, remove features with near-zero
variance, and keep only one instance of perfectly correlated features.
We set the default options based on those implemented in FIDDLE <span class="citation">(Tang et al. 2020)</span>. More details on how to use
<code><a href="../reference/preprocess_data.html">preprocess_data()</a></code> can be found in the accompanying <a href="http://www.schlosslab.org/mikropml/articles/preprocess.html">vignette</a>.</p>
</div>
<div class="section level3">
<h3 id="running-ml">Running ML<a class="anchor" aria-label="anchor" href="#running-ml"></a>
</h3>
<p>The main function in mikropml, <code><a href="../reference/run_ml.html">run_ml()</a></code>, minimally takes
in the model choice and a data frame with an outcome column and feature
columns. For model choice, <code>mikropml</code> currently supports
logistic and linear regression <span class="citation">(<code>glmnet</code>: Friedman, Hastie, and Tibshirani
2010)</span>, support vector machines with a radial basis kernel <span class="citation">(<code>kernlab</code>: Karatzoglou et al. 2004)</span>,
decision trees <span class="citation">(<code>rpart</code>: Therneau et
al. 2019)</span>, random forest <span class="citation">(<code>randomForest</code>: Liaw and Wiener
2002)</span>, and gradient-boosted trees <span class="citation">(<code>xgboost</code>: Chen et al. 2020)</span>.
<code><a href="../reference/run_ml.html">run_ml()</a></code> randomly splits the data into train and test sets
while maintaining the distribution of the outcomes found in the full
dataset. It also provides the option to split the data into train and
test sets based on categorical variables (e.g. batch, geographic
location, etc.). <code>mikropml</code> uses the <code>caret</code>
package <span class="citation">(Kuhn 2008)</span> to train and evaluate
the models, and optionally quantifies feature importance. The output
includes the best model built based on tuning hyperparameters in an
internal and repeated cross-validation step, model evaluation metrics,
and optional feature importances. Feature importances are calculated
using a permutation test, which breaks the relationship between the
feature and the true outcome in the test data, and measures the change
in model performance. This provides an intuitive metric of how
individual features influence model performance and is comparable across
model types, which is particularly useful for model interpretation <span class="citation">(Topçuoğlu et al. 2020)</span>. Our <a href="http://www.schlosslab.org/mikropml/articles/introduction.html">introductory
vignette</a> contains a comprehensive tutorial on how to use
<code><a href="../reference/run_ml.html">run_ml()</a></code>.</p>
<div class="figure">
<img src="mikRopML-pipeline.png" style="width:100.0%" alt=""><p class="caption">mikropml pipeline</p>
</div>
</div>
<div class="section level3">
<h3 id="ideal-workflow-for-running-mikropml-with-many-different-traintest-splits">Ideal workflow for running mikropml with many different train/test
splits<a class="anchor" aria-label="anchor" href="#ideal-workflow-for-running-mikropml-with-many-different-traintest-splits"></a>
</h3>
<p>To investigate the variation in model performance depending on the
train and test set used <span class="citation">(Topçuoğlu et al. 2020;
Lapp et al. 2020)</span>, we provide examples of how to
<code><a href="../reference/run_ml.html">run_ml()</a></code> many times with different train/test splits and
how to get summary information about model performance on <a href="http://www.schlosslab.org/mikropml/articles/parallel.html">a local
computer</a> or on a high-performance computing cluster using a <a href="https://github.com/SchlossLab/mikropml-snakemake-workflow" class="external-link">Snakemake
workflow</a>.</p>
</div>
<div class="section level3">
<h3 id="tuning-visualization">Tuning &amp; visualization<a class="anchor" aria-label="anchor" href="#tuning-visualization"></a>
</h3>
<p>One particularly important aspect of ML is hyperparameter tuning. We
provide a reasonable range of default hyperparameters for each model
type. However practitioners should explore whether that range is
appropriate for their data, or if they should customize the
hyperparameter range. Therefore, we provide a function
<code><a href="../reference/plot_hp_performance.html">plot_hp_performance()</a></code> to plot the cross-validation
performance metric of a single model or models built using different
train/test splits. This helps evaluate if the hyperparameter range is
being searched exhaustively and allows the user to pick the ideal set.
We also provide summary plots of test performance metrics for the many
train/test splits with different models using
<code><a href="../reference/plot_model_performance.html">plot_model_performance()</a></code>. Examples are described in the
accompanying <a href="http://www.schlosslab.org/mikropml/articles/tuning.html">vignette
on hyperparameter tuning</a>.</p>
</div>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<p>mikropml is written in R <span class="citation">(R Core Team
2020)</span> and depends on several packages: <code>dplyr</code> <span class="citation">(Wickham et al. 2020)</span>, <code>rlang</code> <span class="citation">(Henry, Wickham, and RStudio 2020)</span> and
<code>caret</code> <span class="citation">(Kuhn 2008)</span>. The ML
algorithms supported by <code>mikropml</code> require:
<code>glmnet</code> <span class="citation">(Friedman, Hastie, and
Tibshirani 2010)</span>, <code>e1071</code> <span class="citation">(Meyer et al. 2020)</span>, and <code>MLmetrics</code>
<span class="citation">(Yan 2016)</span> for logistic regression,
<code>rpart2</code> <span class="citation">(Therneau et al. 2019)</span>
for decision trees, <code>randomForest</code> <span class="citation">(Liaw and Wiener 2002)</span> for random forest,
<code>xgboost</code> <span class="citation">(Chen et al. 2020)</span>
for xgboost, and <code>kernlab</code> <span class="citation">(Karatzoglou et al. 2004)</span> for support vector
machines. We also allow for parallelization of cross-validation and
other steps using the <code>foreach</code>, <code>doFuture</code>,
<code>future.apply</code>, and <code>future</code> packages <span class="citation">(Bengtsson and Team 2020)</span>. Finally, we use
<code>ggplot2</code> for plotting <span class="citation">(Wickham
2016)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>We thank members of the Schloss Lab who participated in code clubs
related to the initial development of the pipeline, made documentation
improvements, and provided general feedback. We also thank Nick Lesniak
for designing the mikropml logo.</p>
<p>We thank the US Research Software Sustainability Institute (NSF
#1743188) for providing training to KLS at the Winter School in Research
Software Engineering.</p>
</div>
<div class="section level2">
<h2 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h2>
<p>Salary support for PDS came from NIH grant 1R01CA215574. KLS received
support from the NIH Training Program in Bioinformatics (T32 GM070449).
ZL received support from the National Science Foundation Graduate
Research Fellowship Program under Grant No. DGE 1256260. Any opinions,
findings, and conclusions or recommendations expressed in this material
are those of the authors and do not necessarily reflect the views of the
National Science Foundation.</p>
</div>
<div class="section level2">
<h2 id="author-contributions">Author contributions<a class="anchor" aria-label="anchor" href="#author-contributions"></a>
</h2>
<p>BDT, ZL, and KLS contributed equally. Author order among the co-first
authors was determined by time since joining the project.</p>
<p>BDT, ZL, and KLS conceptualized the study and wrote the code. KLS
structured the code in R package form. BDT, ZL, JW, and PDS developed
methodology. PDS, ES, and JW supervised the project. BDT, ZL, and KLS
wrote the original draft. All authors reviewed and edited the
manuscript.</p>
</div>
<div class="section level2">
<h2 id="conflicts-of-interest">Conflicts of interest<a class="anchor" aria-label="anchor" href="#conflicts-of-interest"></a>
</h2>
<p>None.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-bengtsson_futureapply_2020" class="csl-entry">
Bengtsson, Henrik, and R Core Team. 2020. <span>“Future.apply:
<span>Apply Function</span> to <span>Elements</span> in
<span>Parallel</span> Using <span>Futures</span>,”</span> July.
</div>
<div id="ref-breiman_random_2001" class="csl-entry">
Breiman, Leo. 2001. <span>“Random Forests.”</span> <em>Machine
Learning</em> 45 (1): 5–32. <a href="https://doi.org/10.1023/A:1010933404324" class="external-link">https://doi.org/10.1023/A:1010933404324</a>.
</div>
<div id="ref-chen_xgboost_2020" class="csl-entry">
Chen, Tianqi, Tong He, Michael Benesty, Vadim Khotilovich, Yuan Tang,
Hyunsu Cho, Kailong Chen, et al. 2020. <span>“Xgboost: <span>Extreme
Gradient Boosting</span>,”</span> June.
</div>
<div id="ref-fisher_all_2018" class="csl-entry">
Fisher, Aaron, Cynthia Rudin, and Francesca Dominici. 2018. <span>“All
Models Are Wrong, but Many Are Useful: <span>Learning</span> a
Variable’s Importance by Studying an Entire Class of Prediction Models
Simultaneously.”</span>
</div>
<div id="ref-friedman_regularization_2010" class="csl-entry">
Friedman, Jerome H., Trevor Hastie, and Rob Tibshirani. 2010.
<span>“Regularization <span>Paths</span> for <span>Generalized Linear
Models</span> via <span>Coordinate Descent</span>.”</span> <em>Journal
of Statistical Software</em> 33 (1): 1–22. <a href="https://doi.org/10.18637/jss.v033.i01" class="external-link">https://doi.org/10.18637/jss.v033.i01</a>.
</div>
<div id="ref-h2o_platform" class="csl-entry">
H2O.ai. 2020. <em><span>H2O</span>: <span>Scalable</span> Machine
Learning Platform</em>. Manual.
</div>
<div id="ref-hagan_women_2020" class="csl-entry">
Hagan, Ada K., Begüm D. Topçuoğlu, Mia E. Gregory, Hazel A. Barton, and
Patrick D. Schloss. 2020. <span>“Women <span>Are Underrepresented</span>
and <span>Receive Differential Outcomes</span> at <span>ASM
Journals</span>: A <span>Six</span>-<span>Year Retrospective
Analysis</span>.”</span> <em>mBio</em> 11 (6). <a href="https://doi.org/10.1128/mBio.01680-20" class="external-link">https://doi.org/10.1128/mBio.01680-20</a>.
</div>
<div id="ref-henry_rlang_2020" class="csl-entry">
Henry, Lionel, Hadley Wickham, and RStudio. 2020. <span>“Rlang:
<span>Functions</span> for <span>Base Types</span> and <span>Core
R</span> and ’<span>Tidyverse</span>’ <span>Features</span>,”</span>
July.
</div>
<div id="ref-karatzoglou_kernlab_2004" class="csl-entry">
Karatzoglou, Alexandros, Alexandros Smola, Kurt Hornik, and Achim
Zeileis. 2004. <span>“Kernlab - <span>An S4 Package</span> for
<span>Kernel Methods</span> in <span>R</span>.”</span> <em>Journal of
Statistical Software</em> 11 (1): 1–20. <a href="https://doi.org/10.18637/jss.v011.i09" class="external-link">https://doi.org/10.18637/jss.v011.i09</a>.
</div>
<div id="ref-koster_snakemakescalable_2012" class="csl-entry">
Köster, Johannes, and Sven Rahmann. 2012. <span>“Snakemakea Scalable
Bioinformatics Workflow Engine.”</span> <em>Bioinformatics</em> 28 (19):
2520–22. <a href="https://doi.org/10.1093/bioinformatics/bts480" class="external-link">https://doi.org/10.1093/bioinformatics/bts480</a>.
</div>
<div id="ref-kuhn_building_2008" class="csl-entry">
Kuhn, Max. 2008. <span>“Building <span>Predictive Models</span> in
<span>R Using</span> the Caret <span>Package</span>.”</span> <em>Journal
of Statistical Software</em> 28 (1): 1–26. <a href="https://doi.org/10.18637/jss.v028.i05" class="external-link">https://doi.org/10.18637/jss.v028.i05</a>.
</div>
<div id="ref-kuhn_tidymodels_2020" class="csl-entry">
Kuhn, Max, Hadley Wickham, and RStudio. 2020. <span>“Tidymodels:
<span>Easily Install</span> and <span>Load</span> the
’<span>Tidymodels</span>’ <span>Packages</span>,”</span> July.
</div>
<div id="ref-lapp_machine_2020" class="csl-entry">
Lapp, Zena, Jennifer Han, Jenna Wiens, Ellie JC Goldstein, Ebbing
Lautenbach, and Evan Snitkin. 2020. <span>“Machine Learning Models to
Identify Patient and Microbial Genetic Factors Associated with
Carbapenem-Resistant <span>Klebsiella</span> Pneumoniae
Infection.”</span> <em>medRxiv</em>, July, 2020.07.06.20147306. <a href="https://doi.org/10.1101/2020.07.06.20147306" class="external-link">https://doi.org/10.1101/2020.07.06.20147306</a>.
</div>
<div id="ref-liaw_classication_2002" class="csl-entry">
Liaw, Andy, and Matthew Wiener. 2002. <span>“Classification and
<span>Regression</span> by <span class="nocase">randomForest</span>”</span> 2: 5.
</div>
<div id="ref-meyer_e1071_2020" class="csl-entry">
Meyer, David, Evgenia Dimitriadou, Kurt Hornik, Andreas Weingessel,
Friedrich Leisch, Chih-Chung Chang (libsvm C++-code), and Chih-Chen Lin
(libsvm C++-code). 2020. <span>“E1071: <span>Misc Functions</span> of
the <span>Department</span> of <span>Statistics</span>,
<span>Probability Theory Group</span> (<span>Formerly</span>:
<span>E1071</span>), <span>TU Wien</span>.”</span>
</div>
<div id="ref-pedregosa_scikit-learn_2011" class="csl-entry">
Pedregosa, Fabian, Gaël Varoquaux, Alexandre Gramfort, Vincent Michel,
Bertrand Thirion, Olivier Grisel, Mathieu Blondel, et al. 2011.
<span>“Scikit-Learn: <span>Machine Learning</span> in
<span>Python</span>.”</span> <em>Journal of Machine Learning
Research</em> 12 (85): 2825–30.
</div>
<div id="ref-pollard_turning_2019" class="csl-entry">
Pollard, Tom J., Irene Chen, Jenna Wiens, Steven Horng, Danny Wong,
Marzyeh Ghassemi, Heather Mattie, Emily Lindemer, and Trishan Panch.
2019. <span>“Turning the Crank for Machine Learning: Ease, at What
Expense?”</span> <em>The Lancet Digital Health</em> 1 (5): e198–99. <a href="https://doi.org/10.1016/S2589-7500(19)30112-8" class="external-link">https://doi.org/10.1016/S2589-7500(19)30112-8</a>.
</div>
<div id="ref-r_core_team_r_2020" class="csl-entry">
R Core Team. 2020. <span>“R: <span>A Language</span> and
<span>Environment</span> for <span>Statistical Computing</span>.”</span>
</div>
<div id="ref-tang_democratizing_2020" class="csl-entry">
Tang, Shengpu, Parmida Davarmanesh, Yanmeng Song, Danai Koutra, Michael
W. Sjoding, and Jenna Wiens. 2020. <span>“Democratizing <span>EHR</span>
Analyses with <span>FIDDLE</span>: A Flexible Data-Driven Preprocessing
Pipeline for Structured Clinical Data.”</span> <em>J Am Med Inform
Assoc</em>, October. <a href="https://doi.org/10.1093/jamia/ocaa139" class="external-link">https://doi.org/10.1093/jamia/ocaa139</a>.
</div>
<div id="ref-teschendorff_avoiding_2019" class="csl-entry">
Teschendorff, Andrew E. 2019. <span>“Avoiding Common Pitfalls in Machine
Learning Omic Data Science.”</span> <em>Nature Materials</em> 18 (5):
422–27. <a href="https://doi.org/10.1038/s41563-018-0241-z" class="external-link">https://doi.org/10.1038/s41563-018-0241-z</a>.
</div>
<div id="ref-therneau_rpart_2019" class="csl-entry">
Therneau, Terry, Beth Atkinson, Brian Ripley (producer of the initial R.
port, and maintainer 1999-2017). 2019. <span>“Rpart: <span>Recursive
Partitioning</span> and <span>Regression Trees</span>,”</span> April.
</div>
<div id="ref-topcuoglu_framework_2020" class="csl-entry">
Topçuoğlu, Begüm D., Nicholas A. Lesniak, Mack T. Ruffin, Jenna Wiens,
and Patrick D. Schloss. 2020. <span>“A <span>Framework</span> for
<span>Effective Application</span> of <span>Machine Learning</span> to
<span>Microbiome</span>-<span>Based Classification
Problems</span>.”</span> <em>mBio</em> 11 (3). <a href="https://doi.org/10.1128/mBio.00434-20" class="external-link">https://doi.org/10.1128/mBio.00434-20</a>.
</div>
<div id="ref-wickham_ggplot2_2016" class="csl-entry">
Wickham, Hadley. 2016. <em>Ggplot2: <span>Elegant Graphics</span> for
<span>Data Analysis</span></em>. Use <span>R</span>! <span>Cham</span>:
<span>Springer International Publishing</span>. <a href="https://doi.org/10.1007/978-3-319-24277-4" class="external-link">https://doi.org/10.1007/978-3-319-24277-4</a>.
</div>
<div id="ref-wickham_dplyr_2020" class="csl-entry">
Wickham, Hadley, Romain François, Lionel Henry, Kirill Müller, and
RStudio. 2020. <span>“Dplyr: <span>A Grammar</span> of <span>Data
Manipulation</span>,”</span> August.
</div>
<div id="ref-wiens_no_2019" class="csl-entry">
Wiens, Jenna, Suchi Saria, Mark Sendak, Marzyeh Ghassemi, Vincent X.
Liu, Finale Doshi-Velez, Kenneth Jung, et al. 2019. <span>“Do No Harm: A
Roadmap for Responsible Machine Learning for Health Care.”</span>
<em>Nat. Med.</em> 25 (9): 1337–40. <a href="https://doi.org/10.1038/s41591-019-0548-6" class="external-link">https://doi.org/10.1038/s41591-019-0548-6</a>.
</div>
<div id="ref-yan_mlmetrics_2016" class="csl-entry">
Yan, Yachen. 2016. <span>“<span>MLmetrics</span>: <span>Machine Learning
Evaluation Metrics</span>.”</span>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/BTopcuoglu" class="external-link">Begüm Topçuoğlu</a>, <a href="https://github.com/zenalapp" class="external-link">Zena Lapp</a>, <a href="https://github.com/kelly-sovacool" class="external-link">Kelly Sovacool</a>, Evan Snitkin, Jenna Wiens, <a href="https://github.com/pschloss" class="external-link">Patrick Schloss</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
